File: Lintcode 086 - Binary Search Tree Iterator.java
Lintcode 086 - Binary Search Tree Iterator.java
Design an iterator over a binary search tree with the following rules:
•	Elements are visited in ascending order (i.e. an in-order traversal)
•	next() and hasNext() queries run in O(1) time in average.
Example
For the following binary search tree, in-order traversal by using iterator is [1, 6, 10, 11, 12]
   10
 /    \
1      11
 \       \
  6       12
 
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 * Example of iterate a tree:
 * BSTIterator iterator = new BSTIterator(root);
 * while (iterator.hasNext()) {
 *    TreeNode node = iterator.next();
 *    do something for node
 * } 
 */
public class BSTIterator {
    private Stack<TreeNode> stack = new Stack<>();
    private TreeNode curt;
    
    // @param root: The root of binary tree.
    public BSTIterator(TreeNode root) {
        curt = root;
    }

    //@return: True if there has next node, or false
    public boolean hasNext() {
        return (curt != null || !stack.isEmpty());
    }
    
    //@return: return next node
    public TreeNode next() {
        while (curt != null) {
            stack.push(curt);
            curt = curt.left;
        }
        
        curt = stack.pop();
        TreeNode node = curt;
        curt = curt.right;
        
        return node;
    }
}

File: Lintcode 086 - Binary Search Tree Iterator.cs
using System;
using System.Collections;

namespace CSharp
{
    class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode (int val)
        {
            this.val = val;
            this.left = this.right = null;
        }
    }

    class BSTIterator {
        private static Stack stack;
        private TreeNode curr;

        public BSTIterator(TreeNode root)
        {
            this.curr = root;
        }

        public bool hasNext(TreeNode root)
        {
            return(curr != null || stack.Count > 0);
        }

        public moveToNext (TreeNode root)
        {
            while(root != null)
            {
                stack.Push(root);
                root = root.left;
            }

            curr = stack.Pop();
            TreeNode node = curr;
            curr = curr.right;

            return node;
        }
    }
}
